In [2]:
import pandas as pd 
import numpy as np 
%config Completer.use_jedi = False

In [3]:
#Dimension Tables
df_customer = pd.read_csv("dim_customer.csv")
df_delivery_partner = pd.read_csv("dim_delivery_partner_.csv")
df_menu = pd.read_csv("dim_menu_item.csv")
df_restaurant = pd.read_csv("dim_restaurant.csv")

#Fact Tables
df_delivery_performance = pd.read_csv("fact_delivery_performance.csv")
df_order_items = pd.read_csv("fact_order_items.csv")
df_orders = pd.read_csv("fact_orders.csv")
df_ratings = pd.read_csv("fact_ratings.csv")

#loaded the datasets

In [4]:
print("Dimension Tables")
print(f'df_customer: {df_customer.shape}')
print(f'df_delivery_partner:{df_delivery_partner.shape}')
print(f'df_menu: {df_menu.shape}')
print(f'df_restaurant: {df_restaurant.shape}')

print("\nFact Tables")
print(f'df_delivery_performance: {df_delivery_performance.shape}')
print(f'df_order_items: {df_order_items.shape}')
print(f'df_orders: {df_orders.shape}')
print(f'df_ratings: {df_ratings.shape}')

#printed the shapes of all tables

Dimension Tables
df_customer: (107776, 4)
df_delivery_partner:(15000, 7)
df_menu: (342671, 6)
df_restaurant: (19995, 7)

Fact Tables
df_delivery_performance: (149166, 4)
df_order_items: (342994, 8)
df_orders: (149166, 11)
df_ratings: (68842, 7)


In [5]:
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107776 entries, 0 to 107775
Data columns (total 4 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   customer_id          107776 non-null  object
 1   signup_date          107776 non-null  object
 2   city                 107776 non-null  object
 3   acquisition_channel  107776 non-null  object
dtypes: object(4)
memory usage: 3.3+ MB


In [6]:
df_delivery_partner.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   delivery_partner_id  15000 non-null  object 
 1   partner_name         15000 non-null  object 
 2   city                 15000 non-null  object 
 3   vehicle_type         15000 non-null  object 
 4   employment_type      15000 non-null  object 
 5   avg_rating           15000 non-null  float64
 6   is_active            15000 non-null  object 
dtypes: float64(1), object(6)
memory usage: 820.4+ KB


In [7]:
df_delivery_performance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149166 entries, 0 to 149165
Data columns (total 4 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   order_id                     149166 non-null  object 
 1   actual_delivery_time_mins    149166 non-null  int64  
 2   expected_delivery_time_mins  149166 non-null  int64  
 3   distance_km                  149166 non-null  float64
dtypes: float64(1), int64(2), object(1)
memory usage: 4.6+ MB


In [8]:
df_menu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 342671 entries, 0 to 342670
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   menu_item_id   342671 non-null  object 
 1   restaurant_id  342671 non-null  object 
 2   item_name      342671 non-null  object 
 3   category       342671 non-null  object 
 4   is_veg         342671 non-null  object 
 5   price          342671 non-null  float64
dtypes: float64(1), object(5)
memory usage: 15.7+ MB


In [9]:
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 342994 entries, 0 to 342993
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   order_id       342994 non-null  object 
 1   item_id        342994 non-null  object 
 2   menu_item_id   342994 non-null  object 
 3   restaurant_id  342994 non-null  object 
 4   quantity       342994 non-null  int64  
 5   unit_price     342994 non-null  float64
 6   item_discount  342994 non-null  float64
 7   line_total     342994 non-null  float64
dtypes: float64(3), int64(1), object(4)
memory usage: 20.9+ MB


In [10]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149166 entries, 0 to 149165
Data columns (total 11 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             149166 non-null  object 
 1   customer_id          149166 non-null  object 
 2   restaurant_id        149166 non-null  object 
 3   delivery_partner_id  143531 non-null  object 
 4   order_timestamp      149166 non-null  object 
 5   subtotal_amount      149166 non-null  float64
 6   discount_amount      149166 non-null  float64
 7   delivery_fee         149166 non-null  float64
 8   total_amount         149166 non-null  float64
 9   is_cod               149166 non-null  object 
 10  is_cancelled         149166 non-null  object 
dtypes: float64(4), object(7)
memory usage: 12.5+ MB


In [11]:
df_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68842 entries, 0 to 68841
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          68825 non-null  object 
 1   customer_id       68825 non-null  object 
 2   restaurant_id     68825 non-null  object 
 3   rating            68825 non-null  float64
 4   review_text       68825 non-null  object 
 5   review_timestamp  68825 non-null  object 
 6   sentiment_score   68825 non-null  float64
dtypes: float64(2), object(5)
memory usage: 3.7+ MB


In [12]:
df_restaurant.info()
#checked their datatypes as well

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19995 entries, 0 to 19994
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   restaurant_id      19995 non-null  object
 1   restaurant_name    19995 non-null  object
 2   city               19995 non-null  object
 3   cuisine_type       19995 non-null  object
 4   partner_type       19995 non-null  object
 5   avg_prep_time_min  19995 non-null  object
 6   is_active          19995 non-null  object
dtypes: object(7)
memory usage: 1.1+ MB


In [13]:
#checked for null values
df_customer.isnull().sum()

customer_id            0
signup_date            0
city                   0
acquisition_channel    0
dtype: int64

In [14]:
df_delivery_partner.isnull().sum()

delivery_partner_id    0
partner_name           0
city                   0
vehicle_type           0
employment_type        0
avg_rating             0
is_active              0
dtype: int64

In [15]:
df_menu.isnull().sum()

menu_item_id     0
restaurant_id    0
item_name        0
category         0
is_veg           0
price            0
dtype: int64

In [16]:
df_order_items.isnull().sum()

order_id         0
item_id          0
menu_item_id     0
restaurant_id    0
quantity         0
unit_price       0
item_discount    0
line_total       0
dtype: int64

In [17]:
df_orders.isnull().sum()

#found 5635 null values in delivery_partner_id column of df_orders dataframe. Recognised that the missing values for delivery_partner_id are cancelled orders and removed them. However, also observed that many cancelled orders have delivery_partner_id. I have kept them for now.

order_id                  0
customer_id               0
restaurant_id             0
delivery_partner_id    5635
order_timestamp           0
subtotal_amount           0
discount_amount           0
delivery_fee              0
total_amount              0
is_cod                    0
is_cancelled              0
dtype: int64

In [18]:
df_ratings.isnull().sum()

#Found 17 null values in the df_ratings dataframe.

order_id            17
customer_id         17
restaurant_id       17
rating              17
review_text         17
review_timestamp    17
sentiment_score     17
dtype: int64

In [19]:
df_restaurant.isnull().sum()

restaurant_id        0
restaurant_name      0
city                 0
cuisine_type         0
partner_type         0
avg_prep_time_min    0
is_active            0
dtype: int64

In [20]:
#Generated summary statistics.
df_orders.describe()

,subtotal_amount,discount_amount,delivery_fee,total_amount
count,149166.000000,149166.000000,149166.000000,149166.000000
mean,314.373203,18.891589,30.069224,325.550833
std,147.346779,29.801279,10.997816,146.790190
min,0.000000,0.000000,0.000000,0.000000
25%,234.960000,0.000000,24.730000,249.842500
50%,309.070000,0.000000,31.485000,321.700000
75%,396.290000,34.690000,38.240000,406.590000
max,900.000000,222.530000,45.000000,944.910000


In [21]:
df_order_items.describe()

,quantity,unit_price,item_discount,line_total
count,342994.000000,342994.000000,342994.000000,342994.000000
mean,1.998012,86.307888,9.193635,143.814273
std,0.818149,79.286627,17.089856,108.096155
min,1.000000,4.330000,0.000000,5.410000
25%,1.000000,39.350000,0.000000,66.340000
50%,2.000000,62.270000,0.000000,112.785000
75%,3.000000,102.230000,12.580000,190.150000
max,3.000000,900.000000,220.660000,900.000000


In [22]:
df_order_items.describe()

,quantity,unit_price,item_discount,line_total
count,342994.000000,342994.000000,342994.000000,342994.000000
mean,1.998012,86.307888,9.193635,143.814273
std,0.818149,79.286627,17.089856,108.096155
min,1.000000,4.330000,0.000000,5.410000
25%,1.000000,39.350000,0.000000,66.340000
50%,2.000000,62.270000,0.000000,112.785000
75%,3.000000,102.230000,12.580000,190.150000
max,3.000000,900.000000,220.660000,900.000000


In [23]:
df_ratings.describe()

,rating,sentiment_score
count,68825.000000,68825.000000
mean,4.052308,0.526154
std,0.940331,0.470165
min,1.000000,-1.000000
25%,3.800000,0.400000
50%,4.400000,0.700000
75%,4.700000,0.850000
max,5.000000,1.000000


In [24]:
#Check duplicates.
df_customer.duplicated().sum()

np.int64(0)

In [25]:
df_delivery_partner.duplicated().sum()

np.int64(0)

In [26]:
df_menu.duplicated().sum()

np.int64(0)

In [27]:
df_restaurant.duplicated().sum()

np.int64(0)

In [28]:
#Converted date columns into datetime format.
df_customer['signup_date'] = pd.to_datetime(df_customer['signup_date'], dayfirst=True, yearfirst=False, format='mixed')

In [29]:
df_customer.head()

,customer_id,signup_date,city,acquisition_channel
0,CUST000007,2025-03-21,Pune,Organic
1,CUST000008,2025-02-07,Kolkata,Referral
2,CUST000009,2025-02-25,Delhi,Paid
3,CUST000010,2025-01-28,Hyderabad,Paid
4,CUST000011,2025-01-18,Kolkata,Organic


In [30]:
df_orders['order_timestamp'] = pd.to_datetime(df_orders['order_timestamp'], errors="coerce" ,format="mixed")

In [31]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149166 entries, 0 to 149165
Data columns (total 11 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             149166 non-null  object        
 1   customer_id          149166 non-null  object        
 2   restaurant_id        149166 non-null  object        
 3   delivery_partner_id  143531 non-null  object        
 4   order_timestamp      149166 non-null  datetime64[ns]
 5   subtotal_amount      149166 non-null  float64       
 6   discount_amount      149166 non-null  float64       
 7   delivery_fee         149166 non-null  float64       
 8   total_amount         149166 non-null  float64       
 9   is_cod               149166 non-null  object        
 10  is_cancelled         149166 non-null  object        
dtypes: datetime64[ns](1), float64(4), object(6)
memory usage: 12.5+ MB


In [32]:
df_ratings['review_timestamp'] = pd.to_datetime(df_ratings['review_timestamp'], errors="coerce", format="mixed")

In [33]:
df_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68842 entries, 0 to 68841
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   order_id          68825 non-null  object        
 1   customer_id       68825 non-null  object        
 2   restaurant_id     68825 non-null  object        
 3   rating            68825 non-null  float64       
 4   review_text       68825 non-null  object        
 5   review_timestamp  68825 non-null  datetime64[ns]
 6   sentiment_score   68825 non-null  float64       
dtypes: datetime64[ns](1), float64(2), object(4)
memory usage: 3.7+ MB


In [34]:
#removed the null values in df_ratings
df_ratings = df_ratings.dropna()

In [35]:
#Found no nulls
df_ratings.isna().sum()

order_id            0
customer_id         0
restaurant_id       0
rating              0
review_text         0
review_timestamp    0
sentiment_score     0
dtype: int64

<b> Assignment 2: Customer Analysis <b>

In [36]:
#Total customers

Total_customers = df_customer['customer_id'].count()
print(f'Total Customers: {Total_customers}')

Total Customers: 107776


The total customers are 1,07,776

<b> Customers by city <b>

In [37]:
#Customers by city
Customers_by_city = df_customer['customer_id'].groupby(df_customer['city']).count()
print(f'Customers by city: \n{Customers_by_city}')

Customers by city: 
city
Ahmedabad     7512
Bengaluru    30281
Chennai      10823
Delhi        15090
Hyderabad    10755
Kolkata       7466
Mumbai       17317
Pune          8532
Name: customer_id, dtype: int64


<b> Bengaluru has the highest customers while Pune possess the lowest number of customers <b>

<b> Customers by acquisition channel <b>

In [38]:
#Customers by acquisition channel
Customers_by_acquisition_channel = df_customer['customer_id'].groupby(df_customer['acquisition_channel']).count()
print(f'Customers by Acquisition Channel: \n{Customers_by_acquisition_channel}')

Customers by Acquisition Channel: 
acquisition_channel
Organic     55382
Paid        25730
Referral    15377
Social      11287
Name: customer_id, dtype: int64


<b> Organic(55,382) drove most of the customers for Quickbite, followed by Paid(25,730). <b>

<b> Top 5 cities <b>

In [39]:
#Top 5 cities
top_5_cities = df_customer['customer_id'].groupby(df_customer['city']).count().sort_values(ascending = False).head(5)
print(f'Customers by Top 5 Cities: \n{top_5_cities}')

Customers by Top 5 Cities: 
city
Bengaluru    30281
Mumbai       17317
Delhi        15090
Chennai      10823
Hyderabad    10755
Name: customer_id, dtype: int64


<b> Bengaluru(30,281) had the most customers followed by Mumbai, Delhi, Chennai, and Hyderabad. <b>

<B> Monthly signups <b>

In [40]:
#Monthly signups
df_customer['month'] = df_customer['signup_date'].dt.month_name()
Monthly_signups = df_customer['month'].value_counts()
print(f'Monthly signups: {Monthly_signups}')

Monthly signups: month
January      20392
February     16558
December     16536
March        15830
April        12105
May           7380
November      5101
June          4748
July          4646
August        3351
September     1129
Name: count, dtype: int64


<b> January had the highest number of signups followed by February, December, March, and April while September had the lowest signups. <b>

<b> Assignment 3: Restaurant Analysis <b>

In [41]:
#df_restaurant.head()
Total_active_restaurants = df_restaurant['restaurant_id'][df_restaurant['is_active'] == 'Y'].count()
print(f'Total Active Restaurants: {Total_active_restaurants}')

Total Active Restaurants: 18050


In [42]:
#Cuisine distribution
Cuisine_distribution = df_restaurant['restaurant_id'].groupby(df_restaurant['cuisine_type']).count().sort_values(ascending=False)
print(f'Cuisine distribution\n{Cuisine_distribution}')

Cuisine distribution
cuisine_type
North Indian    3942
Biryani         3024
South Indian    3015
Chinese         2970
Pizza           2452
Fast Food       1986
Desserts        1614
Healthy          992
Name: restaurant_id, dtype: int64


<b> North Indian cuisine was the most preferred cuisine by customers, while Healthy cuisine preferred rarely. <b>

In [43]:
#Restaurants by city
Restaurant_by_city = df_restaurant['restaurant_id'].groupby(df_restaurant['city']).count().sort_values(ascending = False)
print(f'Restaurants by City \n{Restaurant_by_city}')

Restaurants by City 
city
Bengaluru    4963
Delhi        2966
Mumbai       2963
Hyderabad    2027
Chennai      2021
Kolkata      1835
Ahmedabad    1630
Pune         1590
Name: restaurant_id, dtype: int64


<b> Bengaluru had the most restaurants whereas Pune had the lowest. <b>

In [44]:
#Active vs inactive
#df_restaurant.pivot_table( index= df_restaurant['city'], columns= df_restaurant['is_active'], aggfunc= ["count"])

Active_inactive = df_restaurant['is_active'].value_counts()
print(f'Active vs Inactive \n{Active_inactive}')

Active vs Inactive 
is_active
Y    18050
N     1945
Name: count, dtype: int64


<b> Active restaurants were more than there were inactive restaurants. <b>

In [45]:
#Average prep time category distribution
df_restaurant['avg_prep_time_min'].groupby(df_restaurant['cuisine_type']).mean

<bound method GroupBy.mean of <pandas.core.groupby.generic.SeriesGroupBy object at 0x00000190A9E5FC70>>

<b> LEVEL 2: Intermediate Pandas <b>

In [46]:
#Gross Revenue
gross_rev = df_orders[df_orders['is_cancelled']=='N']
gross_rev['subtotal_amount'].sum().round(2)

np.float64(46893793.21)

Gross revenue was Rs. 4,68,93,793.

In [47]:
#Net Revenue
net_rev = gross_rev['total_amount'].sum()
net_rev

np.float64(48561115.56)

Net revenue was Rs. 4,85,61,115.

In [48]:
#Total Discounts
total_discounts = gross_rev['discount_amount'].sum()
total_discounts.round(2)

np.float64(2817982.77)

Total Discounts offered was Rs. 28,17,982.

In [49]:
#Delivery Revenue
delivery_rev = gross_rev['delivery_fee'].sum()
delivery_rev

np.float64(4485305.92)

Delivery revenue was Rs. 44,85,305

In [50]:
#AOV (Average Order Value)
aov = gross_rev['subtotal_amount'].sum()/gross_rev['order_id'].count()
aov.round(2)

np.float64(339.68)

Average Order value was Rs. 339.68.

In [51]:
#Revenue by Month
gross_rev.groupby(gross_rev['order_timestamp'].dt.month_name())['subtotal_amount'].sum().sort_values(ascending = False)

order_timestamp
March        7533281.86
January      7512822.09
February     7214790.37
May          7195624.36
April        6882073.23
June         2789338.26
July         2628719.96
September    2586074.13
August       2551068.95
Name: subtotal_amount, dtype: float64

The highest revenue was generated in March while lowest was generated in August.

In [52]:
#Revenue by city
orders_restaurant = pd.merge(left=df_restaurant, right=gross_rev, how= 'inner', on='restaurant_id')
orders_restaurant.groupby(orders_restaurant['city'])['subtotal_amount'].sum().sort_values(ascending = False)

city
Bengaluru    11572173.47
Delhi         6980311.65
Mumbai        6962109.57
Hyderabad     4763909.23
Chennai       4700703.80
Kolkata       4303291.90
Ahmedabad     3869963.26
Pune          3741330.33
Name: subtotal_amount, dtype: float64

The highest revenue was generated in Bengaluru while lowest was generated in Pune city.

In [53]:
#Revenue by Restaurant
orders_restaurant.groupby(['restaurant_id', 'restaurant_name'])['subtotal_amount'].sum().sort_values(ascending = False)

restaurant_id  restaurant_name              
REST16780      Hot & Crispy Pizza Bhojanalay    8170.11
REST13412      Punjabi Darshini Palace          7090.11
REST11694      Punjabi Wraps Bhojanalay         6540.39
REST17743      Hot & Crispy Paratha Zone        6416.13
REST16529      Thindi Mane Grill Palace         6327.20
                                                 ...   
REST17386      Classic Cafe Hub                  186.25
REST00517      Classic Tandoor Zone              184.73
REST18263      Sri House Nest                    172.28
REST02766      Bombay Biryani Bhojanalay         129.49
REST00822      Annapurna Darshini Darbar         121.06
Name: subtotal_amount, Length: 19971, dtype: float64

The Hot & Crispy Pizza Bhojanalay restaurant by restaurant id REST16780 earned the most Rs. 8170.11 while Annapurna Darshini Darbar by restaurant id earned lowest Rs. 121.06 amont 19971 restaurants.


In [54]:
#Revenue by Cuisine
orders_restaurant.groupby(['cuisine_type'])['subtotal_amount'].sum().sort_values(ascending=False)

cuisine_type
North Indian    9270090.55
Biryani         7092170.10
South Indian    7066566.75
Chinese         6922669.86
Pizza           5788392.47
Fast Food       4564169.21
Desserts        3814052.35
Healthy         2375681.92
Name: subtotal_amount, dtype: float64

Among 8 cuisine types, North Indian cuisine generated most revenue Rs. 92,70,090.55 while Healthy food generated the least Rs. 23,75,681.92.

<b> Assignment 5: Restaurant Performance <b>

In [55]:
#Top 10 restaurants by revenue
#orders_restaurant.groupby(['restaurant_id','restaurant_name']).agg({'subtotal_amount': 'sum'}).sort_values(by = 'subtotal_amount',ascending = False).head(5)
orders_restaurant.groupby(['restaurant_id','restaurant_name'])['subtotal_amount'].sum().sort_values(ascending = False).head(10)

restaurant_id  restaurant_name              
REST16780      Hot & Crispy Pizza Bhojanalay    8170.11
REST13412      Punjabi Darshini Palace          7090.11
REST11694      Punjabi Wraps Bhojanalay         6540.39
REST17743      Hot & Crispy Paratha Zone        6416.13
REST16529      Thindi Mane Grill Palace         6327.20
REST15713      Spicy Delights Bhojanalay        6154.08
REST19297      Thindi Mane House Hub            6147.51
REST03377      Fresh Thali Adda                 6139.85
REST07317      Bombay Kitchen Heaven            6120.95
REST18518      Punjabi Curry Delight            6091.78
Name: subtotal_amount, dtype: float64

Top 10 restaurants by revenue were namely Hot & Crispy Pizza Bhojanalay, Punjabi Darshini Palace, Punjabi Wraps Bhojanalay, Hot & Crispy Paratha Zone, Thindi Mane Grill Palace, Spicy Delights Bhojanalay, Thindi Mane House Hub, Fresh Thali Adda, Bombay Kitchen Heaven and Punjabi Curry Delight.

In [56]:
#Top 10 restaurants by orders
orders_restaurant.groupby(['restaurant_id','restaurant_name'])['order_id'].count().sort_values(ascending = False).head(10)

restaurant_id  restaurant_name              
REST16780      Hot & Crispy Pizza Bhojanalay    21
REST16529      Thindi Mane Grill Palace         21
REST13412      Punjabi Darshini Palace          20
REST19297      Thindi Mane House Hub            20
REST11694      Punjabi Wraps Bhojanalay         19
REST03377      Fresh Thali Adda                 19
REST17743      Hot & Crispy Paratha Zone        18
REST12335      Hot & Crispy Express Darbar      18
REST10196      Grand Express Bhojanalay         17
REST03388      Flavours of Delights Darbar      17
Name: order_id, dtype: int64

The top 10 restaurants by orders were Hot & Crispy Pizza Bhojanalay, Thindi Mane Grill Palace, Punjabi Darshini Palace, Thindi Mane House Hub, Punjabi Wraps Bhojanalay, Fresh Thali Adda, Hot & Crispy Paratha Zone, Hot & Crispy Express Darbar, Grand Express Bhojanalay and Flavours of Delights Darbar.

In [57]:
#Top 10 restaurants by AOV
aov_df = orders_restaurant.groupby(['restaurant_id','restaurant_name']).agg(total_rev = ('subtotal_amount', 'sum'), total_orders = ('order_id', 'count')).reset_index()
aov_df['aov'] = aov_df['total_rev']/aov_df['total_orders']
aov_df.sort_values(by='aov',ascending=False).head(10)

,restaurant_id,restaurant_name,total_rev,total_orders,aov
7831,REST07844,Kolkata Grill Clouds,900.00,1,900.000
2842,REST02848,Annapurna Kitchen Point,806.12,1,806.120
10163,REST10179,Fresh Pizza Bhojanalay,739.04,1,739.040
15116,REST15137,Classic Darshini Bhojanalay,719.66,1,719.660
17144,REST17166,Fresh Thali Point,662.44,1,662.440
9048,REST09063,Classic Cafe Bhojanalay,1894.47,3,631.490
8022,REST08035,Thindi Mane Biryani Adda,623.72,1,623.720
17279,REST17301,Bombay Tandoor Bhojanalay,1232.31,2,616.155
16084,REST16105,Fresh Darshini Darbar,613.44,1,613.440
17682,REST17704,Madras Mess Delight,1215.77,2,607.885


In [58]:
#Revenue contribution %
rev_rest = orders_restaurant.groupby(['restaurant_id','restaurant_name'])['subtotal_amount'].sum()
total_rev_rest = rev_rest.sum()
rev_df = (rev_rest*100)/total_rev_rest
rev_df

restaurant_id  restaurant_name       
REST00001      Madras House Central      0.004803
REST00002      Namma Kitchen Clouds      0.006425
REST00003      Taste of Wraps Nest       0.004010
REST00004      Grand Thali Delight       0.003332
REST00005      Taste of Curry Zone       0.010925
                                           ...   
REST19996      Kolkata Kitchen Darbar    0.003962
REST19997      Kolkata Wraps Adda        0.006714
REST19998      Madras Biryani Mahal      0.007316
REST19999      Kolkata Thali Clouds      0.002910
REST20000      Royal House Point         0.003701
Name: subtotal_amount, Length: 19971, dtype: float64

In [59]:
#cancellation Rate
order_rest_all = pd.merge(left=df_orders, right=df_restaurant, on='restaurant_id', how='inner')
order_rest_all['cancel_flag'] = (order_rest_all['is_cancelled']=='Y')
order_rest_all
cancel_rest = order_rest_all.groupby(['restaurant_id','restaurant_name']).agg(total_cancel = ('cancel_flag','sum'), total_orders = ('order_id','count')).reset_index()
cancel_rest['cancellation_rate'] = (cancel_rest['total_cancel']*100)/cancel_rest['total_orders']
cancel_rest

,restaurant_id,restaurant_name,total_cancel,total_orders,cancellation_rate
0,REST00001,Madras House Central,0,6,0.000000
1,REST00002,Namma Kitchen Clouds,1,11,9.090909
2,REST00003,Taste of Wraps Nest,1,8,12.500000
3,REST00004,Grand Thali Delight,1,7,14.285714
4,REST00005,Taste of Curry Zone,1,14,7.142857
...,...,...,...,...,...
19978,REST19996,Kolkata Kitchen Darbar,2,7,28.571429
19979,REST19997,Kolkata Wraps Adda,0,7,0.000000
19980,REST19998,Madras Biryani Mahal,0,9,0.000000
19981,REST19999,Kolkata Thali Clouds,2,8,25.000000


<b> LEVEL 3: Advanced Pandas <b>

Assignment 7: Customer Retention

In [60]:
#Customer Lifespan

cust_orders = pd.merge(left=df_customer, right=df_orders, on='customer_id', how='inner')
#cust_orders.head()
cust_lifespan = cust_orders.groupby(['customer_id']).agg(first_order = ('order_timestamp','min'), last_order = ('order_timestamp','max'))
cust_lifespan['lifespan'] = cust_lifespan['last_order']-cust_lifespan['first_order']
cust_lifespan.sort_values(by='lifespan', ascending=False)

,first_order,last_order,lifespan
customer_id,,,
CUST131883,2025-01-01 22:26:00,2025-09-30 14:04:00,271 days 15:38:00
CUST038375,2025-01-01 13:53:00,2025-09-29 22:16:00,271 days 08:23:00
CUST033709,2025-01-02 12:43:00,2025-09-29 13:00:00,270 days 00:17:00
CUST030720,2025-01-02 14:20:00,2025-09-29 13:23:00,269 days 23:03:00
CUST035922,2025-01-01 19:55:00,2025-09-28 13:31:00,269 days 17:36:00
...,...,...,...
CUST109772,2025-04-29 22:19:00,2025-04-29 22:19:00,0 days 00:00:00
CUST042086,2025-04-17 20:30:00,2025-04-17 20:30:00,0 days 00:00:00
CUST109775,2025-04-06 12:30:00,2025-04-06 12:30:00,0 days 00:00:00


Customer with most lifespan was 271 days while lowest was 0 days.

In [61]:
df_customer.to_csv("dim_customer_cleaned.csv")
df_delivery_partner.to_csv("dim_delivery_partner_cleaned.csv")
df_menu.to_csv("dim_menu_item_cleaned.csv")
df_restaurant.to_csv("dim_restaurant_cleaned.csv")

#Fact Tables
df_delivery_performance.to_csv("fact_delivery_performance_cleaned.csv")
df_order_items.to_csv("fact_order_items_cleaned.csv")
df_orders.to_csv("fact_orders_cleaned.csv")
df_ratings.to_csv("fact_ratings_cleaned.csv")